File: backend/src/main/java/com/main/Difficult/Domain/Factura.java
package com.main.Difficult.Domain;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Getter;
import lombok.Setter;
import org.springframework.data.neo4j.core.schema.Node;
import org.springframework.data.neo4j.core.schema.Relationship;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.Transient;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity @Node("Factura")
public class Factura {
    @org.springframework.data.neo4j.core.schema.Id @org.springframework.data.neo4j.core.schema.GeneratedValue
    @Id @GeneratedValue
    public Long ordenDeComptra = null;
    @Getter @Setter private Integer cantidadDeArticulos = 0;
    @Getter @Setter private Double importeTotal = 0.00;
    @Relationship(type = "COMPRA",direction = Relationship.Direction.INCOMING)
    @Transient @JsonIgnore
    @Getter @Setter private Usuario user = new Usuario();
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    @Getter @Setter private LocalDateTime fechaDeCompra = LocalDateTime.now();

    @Relationship(type = "ITEMS")
    @JsonIgnore
    @Transient
    @Getter @Setter private List<ItemFactura> items = new ArrayList<>();
    private Integer sumOfItemListCantidad(List<ItemFactura> _items){
        Integer sum = 0;
        for (ItemFactura item : _items) {
           sum += item.getCantidad();
        }
        return sum;
    }
    private Double sumOfItemListPrecioCompra(List<ItemFactura> _items){
        double precio = 0;
        for (ItemFactura item : _items) {
            precio += item.getPrecioCompra();
        }
        return precio;
    }

    public Factura(List<ItemFactura> _items, Usuario _user){
        this.setItems(_items);
        this.setUser(_user);
        this.setCantidadDeArticulos(sumOfItemListCantidad(_items));
        this.setImporteTotal(sumOfItemListPrecioCompra(_items));
    }
}

File: backend/src/main/kotlin/aplicacion/dominio/usuario/Factura.kt
package aplicacion.dominio.usuario

import aplicacion.dominio.Item
import aplicacion.dominio.producto.Lote
import aplicacion.dominio.producto.Producto
import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonIgnore
import java.time.LocalDateTime
import javax.persistence.*

@Entity
class Factura() {
    @Id @GeneratedValue(strategy=GenerationType.IDENTITY)
    var ordenDeCompra: Long? = null
    var cantidadDeArticulos: Int = 0
    var importeTotal: Double = 0.00

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    var fechaDeCompra: LocalDateTime = LocalDateTime.now()

    @OneToMany(cascade = [CascadeType.MERGE,CascadeType.PERSIST]) @JoinColumn(name="factura_id")
    @JsonIgnore lateinit var items: List<ItemFactura>

    constructor(items: List<ItemFactura>) : this() {
        this.items = items

        this.cantidadDeArticulos = items.size
        this.importeTotal = items.sumOf { it.precioCompra }
    }
}

@Entity @Table(name="factura_producto_lote")
class ItemFactura(): Item() {
    var precioCompra: Double = 0.0
    constructor(producto: Producto, lote: Lote, cantidad: Int) : this() {
        this.producto = producto
        this.lote = lote
        this.cantidad = cantidad

        lote.descontar(cantidad)
        precioCompra = cantidad * producto.precio()
    }
}